File: python/src/puff1.py
#!/usr/bin/env python3
# Copyright (C) 2026 Lenik <zephyr@bodz.net>
# SPDX-License-Identifier: AGPL-3.0-or-later

from __future__ import annotations

import os
import sys
from typing import TextIO

from pufflib import copy_file, copy_stream, init_i18n


def usage(out: TextIO) -> None:
    out.write(
        _("Usage: puff1 [OPTION]... [FILE]...\n"
          "Concatenate FILE(s) to standard output. With no FILE, or when FILE is -,\n"
          "read standard input.\n")
    )
    out.write("\n")
    out.write("  -v, --verbose      ")
    out.write(_("repeat for more verbose loggings\n"))
    out.write("  -q, --quiet        ")
    out.write(_("show less logging messages\n"))
    out.write("  -h, --help         ")
    out.write(_("display this help and exit\n"))
    out.write("      --version      ")
    out.write(_("output version information and exit\n"))
    out.write("\n")
    out.write(_("Report bugs to: <{email}>\n").format(email="zephyr@bodz.net"))


def version(out: TextIO) -> None:
    out.write("puff1 dev\n")
    out.write(_("Copyright (C) {year} {author}\n").format(year=2026, author="Lenik"))
    out.write(_("License AGPL-3.0-or-later: <https://www.gnu.org/licenses/agpl-3.0.html>\n"))
    out.write(_("This is free software: you are free to change and redistribute it.\n"))
    out.write(_("This project opposes AI exploitation and AI hegemony.\n"))
    out.write(
        _(
            "This project rejects mindless MIT-style licensing and politically naive "
            "BSD-style licensing.\n"
        )
    )
    out.write(_("There is NO WARRANTY, to the extent permitted by law.\n"))


def main(argv: list[str]) -> int:
    init_i18n(argv[0])

    args = argv[1:]
    verbose = 0
    files: list[str] = []
    i = 0
    while i < len(args):
        a = args[i]
        if a in ("-h", "--help"):
            usage(sys.stdout)
            return 0
        if a == "--version":
            version(sys.stdout)
            return 0
        if a in ("-v", "--verbose"):
            verbose += 1
            i += 1
            continue
        if a in ("-q", "--quiet"):
            verbose = -1
            i += 1
            continue
        files.append(a)
        i += 1

    if verbose > 0:
        print(f"{argv[0]}: verbose mode enabled", file=sys.stderr)

    out = sys.stdout.buffer
    if not files:
        copy_stream(sys.stdin.buffer, out)
        return 0

    for p in files:
        if p == "-":
            copy_stream(sys.stdin.buffer, out)
        else:
            try:
                copy_file(p, out)
            except OSError as e:
                print(f"{argv[0]}: {e}", file=sys.stderr)
                return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: erlang/src/puff1.erl
-module(puff1).
-export([main/0]).

tr(S) -> S.

usage() ->
    io:put_chars(tr("Usage: puff1 [OPTION]... [FILE]...\n")),
    io:put_chars(tr("Concatenate FILE(s) to standard output. With no FILE, or when FILE is -,\n")),
    io:put_chars(tr("read standard input.\n\n")),
    io:put_chars("  -v, --verbose      "),
    io:put_chars(tr("repeat for more verbose loggings\n")),
    io:put_chars("  -q, --quiet        "),
    io:put_chars(tr("show less logging messages\n")),
    io:put_chars("  -h, --help         "),
    io:put_chars(tr("display this help and exit\n")),
    io:put_chars("      --version      "),
    io:put_chars(tr("output version information and exit\n\n")),
    io:format(tr("Report bugs to: <~s>~n"), ["zephyr@bodz.net"]).

version_info() ->
    io:put_chars("puff1 dev\n"),
    io:put_chars(tr("Copyright (C) 2026 Lenik\n")),
    io:put_chars(tr("License AGPL-3.0-or-later: <https://www.gnu.org/licenses/agpl-3.0.html>\n")),
    io:put_chars(tr("This is free software: you are free to change and redistribute it.\n")),
    io:put_chars(tr("This project opposes AI exploitation and AI hegemony.\n")),
    io:put_chars(tr("This project rejects mindless MIT-style licensing and politically naive BSD-style licensing.\n")),
    io:put_chars(tr("There is NO WARRANTY, to the extent permitted by law.\n")).

copy_files([]) ->
    ok;
copy_files(["-" | Rest]) ->
    case common_lib:copy_stream(standard_io, standard_io) of
        ok -> copy_files(Rest);
        {error, Reason} -> {error, Reason}
    end;
copy_files([File | Rest]) ->
    case common_lib:copy_file(File) of
        ok -> copy_files(Rest);
        {error, Reason} -> {error, {File, Reason}}
    end.

main() ->
    Args = init:get_plain_arguments(),
    case lists:member("-h", Args) orelse lists:member("--help", Args) of
        true ->
            usage(),
            erlang:halt(0);
        false ->
            ok
    end,
    case lists:member("--version", Args) of
        true ->
            version_info(),
            erlang:halt(0);
        false ->
            ok
    end,
    Verbose = lists:member("-v", Args) orelse lists:member("--verbose", Args),
    Files = [A || A <- Args, not lists:member(A, ["-v", "--verbose", "-q", "--quiet"])],
    case Verbose of
        true -> io:format(standard_error, "~s: verbose mode enabled~n", ["puff1"]);
        false -> ok
    end,
    case Files of
        [] ->
            _ = common_lib:copy_stream(standard_io, standard_io),
            erlang:halt(0);
        _ ->
            case copy_files(Files) of
                ok ->
                    erlang:halt(0);
                {error, {File, Reason}} ->
                    io:format(standard_error, "puff1: ~s: ~p~n", [File, Reason]),
                    erlang:halt(1);
                {error, Reason} ->
                    io:format(standard_error, "puff1: ~p~n", [Reason]),
                    erlang:halt(1)
            end
    end.
